src/GuzzleRequestAdapter.php

```diff
@@ -281,152 +356,282 @@ public function getBaseUrl(): string
+                $matches       = [];
+                if (!preg_match(self::$claimsRegex, $wwwAuthHeader, $matches)) {
                     return $response;
                 }
-                $request->content->rewind();
-            }
-            $wwwAuthHeader = $response->getHeaderLine(self::$wwwAuthenticateHeader);
-            $matches = [];
-            if (!preg_match(self::$claimsRegex, $wwwAuthHeader, $matches)) {
-                return $response;
+                $claims = $matches[1];
+                /** @var ResponseInterface $response */
+                $response = $this->getHttpResponseMessage($requestInformation, $claims)->wait();
+            } finally {
+                $span->end();
             }
-            $claims = $matches[1];
-            /** @var ResponseInterface $response */
-            $response =  $this->getHttpResponseMessage($request, $claims)->wait();
         }
         return $response;
     }
 
+    public const ERROR_BODY_FOUND_ATTRIBUTE_NAME = "com.microsoft.kiota.error.body_found";
+    public const ERROR_MAPPING_FOUND_ATTRIBUTE_NAME = "com.microsoft.kiota.error.mapping_found";
+
     /**
      * @template T of Parsable
      * @param ResponseInterface $response
      * @param array<string, array{class-string<T>, string}>|null $errorMappings
+     * @param SpanInterface $span
      * @throws ApiException
      */
-    private function throwFailedResponse(ResponseInterface $response, ?array $errorMappings): void {
+    private function throwFailedResponse(ResponseInterface $response, ?array $errorMappings, SpanInterface $span): void {
+        $currentContext = Context::getCurrent();
         $statusCode = $response->getStatusCode();
         if ($statusCode >= 200 && $statusCode < 400) {
             return;
         }
-        $statusCodeAsString = "$statusCode";
-        if ($errorMappings === null || (!array_key_exists($statusCodeAsString, $errorMappings) &&
-            !($statusCode >= 400 && $statusCode < 500 && isset($errorMappings['4XX'])) &&
-            !($statusCode >= 500 && $statusCode < 600 && isset($errorMappings["5XX"])))) {
-            $ex = new ApiException("the server returned an unexpected status code and no error class is registered for this code " . $statusCode);
-            $ex->setResponseStatusCode($response->getStatusCode());
-            $ex->setResponseHeaders($response->getHeaders());
-            throw $ex;
-        }
-        $errorClass = array_key_exists($statusCodeAsString, $errorMappings) ? $errorMappings[$statusCodeAsString] : ($errorMappings[$statusCodeAsString[0] . 'XX'] ?? null);
+        $errorSpan = $this->tracer->spanBuilder('throwFailedResponse')
+            ->setParent($currentContext)
+            ->startSpan();
+        $scope = $errorSpan->activate();
+        try {
+            $responseBodyContents = $response->getBody()->getContents();
+
+            $errorSpan->setStatus(StatusCode::STATUS_ERROR, 'received_error_response');
```

Comment:
We need to also set an attribute "status_message" to the value "received_error_response" down the various blocks where an error is processed.